plugins/processors/awsentity/processor.go

```diff
@@ -0,0 +1,413 @@
+
+				ec2Attributes := EC2ServiceAttributes{
+					InstanceId:        ec2Info.InstanceID,
+					AutoScalingGroup:  ec2Info.AutoScalingGroup,
+					ServiceNameSource: entityServiceNameSource,
+				}
+				if err := validate.Struct(ec2Attributes); err == nil {
+					resourceAttrs.PutStr(entityattributes.AttributeEntityPlatformType, entityPlatformType)
+					AddAttributeIfNonEmpty(resourceAttrs, entityattributes.AttributeEntityInstanceID, ec2Attributes.InstanceId)
+					AddAttributeIfNonEmpty(resourceAttrs, entityattributes.AttributeEntityAutoScalingGroup, ec2Attributes.AutoScalingGroup)
+					AddAttributeIfNonEmpty(resourceAttrs, entityattributes.AttributeEntityServiceNameSource, ec2Attributes.ServiceNameSource)
+				}
+			}
+			if logGroupNames == EMPTY || (serviceName == EMPTY && environmentName == EMPTY) {
+				continue
+			}
+
+			logGroupNamesSlice := strings.Split(logGroupNames, "&")
+			for _, logGroupName := range logGroupNamesSlice {
+				if logGroupName == EMPTY {
+					continue
+				}
+				addToEntityStore(entitystore.LogGroupName(logGroupName), serviceName, environmentName)
+			}
+		}
+
+	}
+	return md, nil
+}
+
+// scrapeServiceAttribute expands the datapoint attributes and search for
+// service name and environment attributes. This is only used for components
+// that only emit attributes on datapoint level.
+func (p *awsEntityProcessor) scrapeServiceAttribute(scopeMetric pmetric.ScopeMetricsSlice) (string, string, string) {
+	entityServiceName := EMPTY
+	entityServiceNameSource := EMPTY
+	entityEnvironmentName := EMPTY
+	for j := 0; j < scopeMetric.Len(); j++ {
+		metric := scopeMetric.At(j).Metrics()
+		for k := 0; k < metric.Len(); k++ {
+			if entityServiceName != EMPTY && entityEnvironmentName != EMPTY && entityServiceNameSource != EMPTY {
+				return entityServiceName, entityEnvironmentName, entityServiceNameSource
+			}
+			m := metric.At(k)
+			switch m.Type() {
+			case pmetric.MetricTypeGauge:
+				dps := m.Gauge().DataPoints()
+				for l := 0; l < dps.Len(); l++ {
+					dpService := getServiceAttributes(dps.At(l).Attributes())
+					if dpService != EMPTY {
+						entityServiceName = dpService
+					}
+					if dpServiceNameSource, ok := dps.At(l).Attributes().Get(entityattributes.AttributeServiceNameSource); ok {
+						entityServiceNameSource = dpServiceNameSource.Str()
+						dps.At(l).Attributes().Remove(semconv.AttributeServiceName)
+						dps.At(l).Attributes().Remove(entityattributes.AttributeServiceNameSource)
+					}
+					if dpEnvironment, ok := dps.At(l).Attributes().Get(semconv.AttributeDeploymentEnvironment); ok {
+						entityEnvironmentName = dpEnvironment.Str()
+					}
+					if _, ok := dps.At(l).Attributes().Get(entityattributes.AttributeDeploymentEnvironmentSource); ok {
+						dps.At(l).Attributes().Remove(semconv.AttributeDeploymentEnvironment)
+						dps.At(l).Attributes().Remove(entityattributes.AttributeDeploymentEnvironmentSource)
+					}
```

Comment:
nit: These blocks can be moved to a function that takes in the `pcommon.Map` of the attributes.

Reply:
Non-blocking comment